U/danielsf/pre screen

danielsf opened the pull request:
while testing the Twinkles catalog generation code, I noticed that CompoundInstanceCatalogs were running much slower than straight concatenations of InstanceCatalogs in a way such that t_compound/t_instance scaled with the size of the catalog.  This was because, when made a part of a CompoundInstanceCatalog, InstanceCatalogs were being forced to deal with rows that they didn't have to ordinarily (i.e. galaxies without AGN being passed to the AGN catalog).  This pull request fixes that by allowing InstanceCatalogs to filter the results of database queries directly using their cannot_be_null list of columns.  CompoundInstanceCatalogs are still 20% slower than a concatenation of InstanceCatalogs, but that ratio does not scale with the size of the catalog (before, an InstanceCatalog drawn on a 0.6 degree patch of sky would be 50% slower when generated with CompoundInstanceCatalog).

There is a related pull request on sims_catUtils

danielsf committed:
added _pre_screen to InstanceCatalog

this is a flag which, when True, causes InstanceCatalog.write_catalog()
to check the results of the database query for any rows that violate
self.cannot_be_null *before* proceeding to calculate getter columns.

danielsf committed:
InstanceCatalogs created by CompoundInstanceCatalog use _pre_screen=True

danielsf committed:
test that cannot_be_null work when several columns are specified

danielsf committed:
rename counters in testInstanceCatalog.py to be more illustrative

jbkalmbach commented on python/lsst/sims/catalogs/definitions/InstanceCatalog.py:
```diff
@@ -163,6 +163,8 @@ class InstanceCatalog(object):
     delimiter = ", "
     comment_char = "#"
     endline = "\n"
+    _pre_screen = False  # if true, write_catalog() will check database query resutls against
```
typo results

jbkalmbach commented on tests/testInstanceCatalog.py:
```diff
@@ -126,6 +126,14 @@ class unicodeCannotBeNullCatalog(InstanceCatalog):
 
+class severalCannotBeNullCatalog(InstanceCatalog):
+    """
+    This catalog class will not write rows with a null value in the n2 column
```
"in n2 or n4", right? Or am I misunderstanding?

jbkalmbach replied:
I also think the doc strings for the cannotBeNull Catalogs above must be corrected in the same way.

jbkalmbach commented on tests/testInstanceCatalog.py:
```diff
@@ -367,20 +376,93 @@ def testCannotBeNull(self):
+        def testCannotBeNull_pre_screen(self):
+            """
+            Test to make sure that the code for filtering out rows with null values
+            in key rowss works.  This time, check that the code works when we set
```
typo rows

jbkalmbach commented on tests/testInstanceCatalog.py:
```diff
@@ -367,20 +376,93 @@ def testCannotBeNull(self):
+            if os.path.exists(fileName):
+                os.unlink(fileName)
+
+        def testCannotBeNull_pre_screen(self):
```
I'm not quite clear on how this test makes sure that self._pre_screen is actually filtering the rows and we're not just using what we have in place already. It seems to me that if you just remove the cat._pre_screen = True line you would get the same results in this test.

danielsf replied:
I have changed test so that it both verifies that the catalogs produced with _pre_screen=True and _pre_screen=False are identical and that the size of _current_chunk is smaller with _pre_screen=True.

danielsf committed:
fix typo in comment

danielsf committed:
fix typos in docstrings

danielsf committed:
make sure the pre-screened catalogs do, in fact, pass smaller chunks

danielsf committed:
simplify pre_screen unit test

since we have already tested the general behavior of cannot_be_null,
all we have to do now is make sure that self._pre_screen = True
does not change the output catalog.

danielsf committed:
make sure all catalogs written to tests/scratchSpace

in testInstanceCatalogs.py

danielsf committed:
fix line that was too long (comment)

danielsf committed:
added severalCannotBeNullCatalog to testCannotBeNull

danielsf committed:
make sure that multiple entries in cannot_be_null works

danielsf committed:
pep8 changes to InstanceCatalog.py

danielsf committed:
pep8 change that rearranged lambda function in InstanceCatalogMeta

danielsf committed:
remove unused variables

danielsf committed:
pep8 changes to CompoundCatalogDBObject.py

danielsf committed:
remove unused variables from CompoundCatalogDBObject.py

danielsf committed:
import numpy as np in CompoundInstanceCatalog.py

danielsf committed:
import numpy as np in InstanceCatalog.py

jbkalmbach commented on tests/testInstanceCatalog.py:
```diff
@@ -388,19 +408,26 @@ def testCannotBeNull(self):
-            InstanceCatalog._pre_screen = True, causing write_catalog to screen
-            the database query results before calculating getter columns.
+            Check that writing a catalog with self._pre_screen = True produces
+            the same results as writing one with self._pre_creen = False, except
```
typo self._pre_screen

danielsf committed:
fix typo
